### PR TITLE
CodeCamp2023-555

### PR DESCRIPTION
**Pull request by Boomerl**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily receiving feedbacks. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a dependency on downstream projects of a newer version, this PR should be tested with all supported versions of downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: tests/regression/mmdet.yml [hunk 1]
```diff
@@ -446,3 +446,13 @@ models:
     pipelines:
       - *pipeline_ort_dynamic_fp32
       - *pipeline_trt_dynamic_fp32
+
+  - name: CondInst
+    metafile: /configs/condinst/metafile.yml
```
Comment (RunningLeon) [inline tests/regression/mmdet.yml:451]: `/configs/condinst/metafile.yml` -> `configs/condinst/metafile.yml`

**Comment by Boomerl** [comment]
10/07 17:11:54 - mmengine - INFO - **********Environmental information**********
10/07 17:11:55 - mmengine - INFO - sys.platform: linux
10/07 17:11:55 - mmengine - INFO - Python: 3.9.17 (main, Jul  5 2023, 20:41:20) [GCC 11.2.0]
10/07 17:11:55 - mmengine - INFO - CUDA available: True
10/07 17:11:55 - mmengine - INFO - numpy_random_seed: 2147483648
10/07 17:11:55 - mmengine - INFO - GPU 0,1,2,3,4,5,6,7: NVIDIA GeForce RTX 3080
10/07 17:11:55 - mmengine - INFO - CUDA_HOME: None
10/07 17:11:55 - mmengine - INFO - GCC: gcc (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
10/07 17:11:55 - mmengine - INFO - PyTorch: 1.12.1+cu116
10/07 17:11:55 - mmengine - INFO - PyTorch compiling details: PyTorch built with:
  - GCC 9.3
  - C++ Version: 201402
  - Intel(R) Math Kernel Library Version 2020.0.0 Product Build 20191122 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v2.6.0 (Git Hash 52b5f107dd9cf10910aaa19cb47f3abf9b349815)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - LAPACK is enabled (usually provided by MKL)
  - NNPACK is enabled
  - CPU capability usage: AVX2
  - CUDA Runtime 11.6
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_80,code=sm_80;-gencode;arch=compute_86,code=sm_86
  - CuDNN 8.6  (built against CUDA 11.8)
    - Built with CuDNN 8.3.2
  - Magma 2.6.1
  - Build settings: BLAS_INFO=mkl, BUILD_TYPE=Release, CUDA_VERSION=11.6, CUDNN_VERSION=8.3.2, CXX_COMPILER=/opt/rh/devtoolset-9/root/usr/bin/c++, CXX_FLAGS= -fabi-version=11 -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_KINETO -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -DEDGE_PROFILER_USE_KINETO -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-unused-parameter -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Werror=cast-function-type -Wno-stringop-overflow, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_VERSION=1.12.1, USE_CUDA=ON, USE_CUDNN=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=OFF, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, USE_ROCM=OFF, 

10/07 17:11:55 - mmengine - INFO - TorchVision: 0.13.1+cu116
10/07 17:11:55 - mmengine - INFO - OpenCV: 4.8.0
10/07 17:11:55 - mmengine - INFO - MMEngine: 0.8.2
10/07 17:11:55 - mmengine - INFO - MMCV: 2.0.1
10/07 17:11:55 - mmengine - INFO - MMCV Compiler: GCC 9.3
10/07 17:11:55 - mmengine - INFO - MMCV CUDA Compiler: 11.6
10/07 17:11:55 - mmengine - INFO - MMDeploy: 1.3.0+be7042e
10/07 17:11:55 - mmengine - INFO - 

10/07 17:11:55 - mmengine - INFO - **********Backend information**********
10/07 17:11:55 - mmengine - INFO - tensorrt:    8.5.1.7
10/07 17:11:55 - mmengine - INFO - tensorrt custom ops: NotAvailable
10/07 17:11:55 - mmengine - INFO - ONNXRuntime: None
10/07 17:11:55 - mmengine - INFO - ONNXRuntime-gpu:     1.10.0
10/07 17:11:55 - mmengine - INFO - ONNXRuntime custom ops:      NotAvailable
10/07 17:11:55 - mmengine - INFO - pplnn:       None
10/07 17:11:55 - mmengine - INFO - ncnn:        None
10/07 17:11:55 - mmengine - INFO - snpe:        None
10/07 17:11:55 - mmengine - INFO - openvino:    None
10/07 17:11:55 - mmengine - INFO - torchscript: 1.12.1+cu116
10/07 17:11:55 - mmengine - INFO - torchscript custom ops:      NotAvailable
10/07 17:11:55 - mmengine - INFO - rknn-toolkit:        None
10/07 17:11:55 - mmengine - INFO - rknn-toolkit2:       None
10/07 17:11:55 - mmengine - INFO - ascend:      None
10/07 17:11:55 - mmengine - INFO - coreml:      None
10/07 17:11:55 - mmengine - INFO - tvm: None
10/07 17:11:55 - mmengine - INFO - vacc:        None
10/07 17:11:55 - mmengine - INFO - 

10/07 17:11:55 - mmengine - INFO - **********Codebase information**********
10/07 17:11:55 - mmengine - INFO - mmdet:       3.0.0
10/07 17:11:55 - mmengine - INFO - mmseg:       None
10/07 17:11:55 - mmengine - INFO - mmpretrain:  None
10/07 17:11:55 - mmengine - INFO - mmocr:       None
10/07 17:11:55 - mmengine - INFO - mmagic:      None
10/07 17:11:55 - mmengine - INFO - mmdet3d:     None
10/07 17:11:55 - mmengine - INFO - mmpose:      None
10/07 17:11:55 - mmengine - INFO - mmrotate:    None
10/07 17:11:55 - mmengine - INFO - mmaction:    None
10/07 17:11:55 - mmengine - INFO - mmrazor:     None

**Comment by RunningLeon** [comment]
@Boomerl hi, pls fix lint error by running `pip install pre-commit && pre-commit run --all-files` inside mmdeploy root directory.

File: tests/regression/mmdet.yml (hunk 1 above)
Comment (RunningLeon) [reply tests/regression/mmdet.yml:451]: @Boomerl  cc

Comment (Boomerl) [reply tests/regression/mmdet.yml]: updated

File: mmdeploy/codebase/mmdet/models/dense_heads/condinst_head.py [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+    locations = self.prior_generator.single_level_grid_priors(
+        hw, level_idx=0, device=mask_feats.device)
+    locations = locations.unsqueeze(0).repeat(batch_size, 1,
+                                              1).reshape(batch_size, 1, -1, 2)
+    centers = points.reshape(batch_size, -1, 1, 2)
+    rel_coordinates = (centers - locations).permute(0, 1, 3, 2).float()
+    rel_coordinates /= (strides[:, :, None, None] * self.size_of_interest)
+    rel_coords = rel_coordinates.reshape(batch_size, -1, 2, hw[0], hw[1])
+    mask_head_inputs = torch.cat([rel_coords, mask_feats], dim=2)
+    # TODO: change following code to support batch inference
```
Comment (RunningLeon) [inline mmdeploy/codebase/mmdet/models/dense_heads/condinst_head.py:135]: could remove this line

File: tests/test_codebase/test_mmdet/test_mmdet_models.py [hunk 3]
```diff
@@ -2364,3 +2364,251 @@ def test_solov2_head_predict_by_feat(backend_type):
...
+        test_cfg=test_cfg,
+    )
+
+    model.requires_grad_(False)
+    return model
+
+
+@pytest.mark.skipif(
+    reason='Only support GPU test', condition=not torch.cuda.is_available())
+@pytest.mark.parametrize('backend_type',
```
Comment (RunningLeon) [inline tests/test_codebase/test_mmdet/test_mmdet_models.py:2547]: we could remove test for tensorrt backend

File: tests/test_codebase/test_mmdet/test_mmdet_models.py [hunk 4]
```diff
@@ -2364,3 +2364,251 @@ def test_solov2_head_predict_by_feat(backend_type):
...
+            type='DiceLoss',
+            use_sigmoid=True),
+        test_cfg=test_cfg,
+    )
+
+    model.requires_grad_(False)
+    return model
+
+
+@pytest.mark.skipif(
```
Comment (RunningLeon) [inline tests/test_codebase/test_mmdet/test_mmdet_models.py:2545]: could test with onnxruntime on cpu

File: tests/test_codebase/test_mmdet/test_mmdet_models.py [hunk 5]
```diff
@@ -2364,3 +2364,251 @@ def test_solov2_head_predict_by_feat(backend_type):
...
+    reason='Only support GPU test', condition=not torch.cuda.is_available())
+@pytest.mark.parametrize('backend_type',
```
Comment (RunningLeon) [inline tests/test_codebase/test_mmdet/test_mmdet_models.py:2416]: could remove test for tensorrt

File: tests/test_codebase/test_mmdet/test_mmdet_models.py [hunk 6]
```diff
@@ -2364,3 +2364,251 @@ def test_solov2_head_predict_by_feat(backend_type):
...
+        'img_shape': (s, s, 3)
+    }]
+
+    output_names = ['dets', 'labels', 'masks']
+    deploy_cfg = Config(
+        dict(
+            backend_config=dict(
+                type=backend_type.value,
+                common_config=dict(max_workspace_size=1 << 32),
+                model_inputs=[
```
Comment (RunningLeon) [inline tests/test_codebase/test_mmdet/test_mmdet_models.py:2565]: could remove these input shapes setting and they are not correct for inputs like:`mask_preds, det, label, batch_img_metas`

File: tests/test_codebase/test_mmdet/test_mmdet_models.py [hunk 7]
```diff
@@ -2364,3 +2364,251 @@ def test_solov2_head_predict_by_feat(backend_type):
...
+        'img_shape': (s, s, 3)
+    }]
+
+    output_names = ['dets', 'labels', 'param_preds', 'points', 'strides']
+    deploy_cfg = Config(
+        dict(
+            backend_config=dict(
+                type=backend_type.value,
+                common_config=dict(max_workspace_size=1 << 32),
+                model_inputs=[
```
Comment (RunningLeon) [inline tests/test_codebase/test_mmdet/test_mmdet_models.py:2436]: could remove these input shapes setting and they are not correct for inputs like:`cls_scores, bbox_preds, label, ...`

**Comment by irexyc** [comment]
--dump-info 模型导出的信息中后处理部分不对，应该是 `ResizeInstanceMask`

这个地方有个bug，resize_height / resize_width 写反了，可以一块改一下。
https://github.com/open-mmlab/mmdeploy/blob/main/csrc/mmdeploy/codebase/mmdet/instance_segmentation.cpp#L152

File: tests/test_codebase/test_mmdet/test_mmdet_models.py [hunk 8]
```diff
@@ -2364,3 +2364,225 @@ def test_solov2_head_predict_by_feat(backend_type):
...
+            onnx_config=dict(output_names=output_names, input_shape=None),
+            codebase_config=dict(type='mmdet', task='ObjectDetection')))
+
+    class TestCondInstMaskHeadModel(torch.nn.Module):
+
+        def __init__(self, condinst_mask_head):
+            super(TestCondInstMaskHeadModel, self).__init__()
+            self.mask_head = condinst_mask_head
+
+        def predict_by_feat(self, mask_preds, det, label, batch_img_metas):
```
Comment (RunningLeon) [inline tests/test_codebase/test_mmdet/test_mmdet_models.py:2556]: change to `forward`

File: tests/test_codebase/test_mmdet/test_mmdet_models.py [hunk 9]
```diff
@@ -2364,3 +2364,225 @@ def test_solov2_head_predict_by_feat(backend_type):
...
+
+    seed_everything(1234)
+    mask_preds = torch.rand(1, 100, 200, 200)
+    seed_everything(5678)
+    dets = torch.rand(1, 100, 5)
+    labels = torch.rand(1, 100)
+
+    # to get outputs of onnx model after rewrite
+    wrapped_model = WrapModel(
+        condinst_mask_head, 'predict_by_feat', batch_img_metas=batch_img_metas)
```
Comment (RunningLeon) [inline tests/test_codebase/test_mmdet/test_mmdet_models.py:2573]: change to `forward`

**Review by RunningLeon — APPROVED** [review]
LGTM

**Comment by RunningLeon** [comment]
@Boomerl hi, pls. add ut and update docs like https://github.com/open-mmlab/mmdeploy/blob/main/docs/en/04-supported-codebases/mmdet.md#supported-models in another PR.

**Comment by Boomerl** [comment]
ok, i will do it soon.
